src/backend/distributed/planner/multi_planner_utils.c

```diff
@@ -0,0 +1,863 @@
+ * A, B and C.
+ *
+ * Note that equality among the members are identified by the varattno and rteIdentity.
+ */
+typedef struct AttributeEquivalenceClass
+{
+	uint32 equivalenceId;
+	List *equivalentAttributes;
+} AttributeEquivalenceClass;
+
+/*
+ *  AttributeEquivalenceClassMember - one member expression of an
+ *  AttributeEquivalenceClass. The important thing to consider is that
+ *  the class member contains "rteIndentity" field. Note that each RTE_RELATION
+ *  is assigned a unique rteIdentity in AssignRTEIdentities() function.
+ *
+ *  "varno" and "varattno" is directly used from a Var clause that is being added
+ *  to the attribute equivalence. Since we only use this class for relations, the member
+ *  also includes the relation id field.
+ */
+typedef struct AttributeEquivalenceClassMember
+{
+	Oid relationId;
+	int rteIdentity;
+	Index varno;
+	AttrNumber varattno;
+} AttributeEquivalenceClassMember;
+
+
+static uint32 ReferenceRelationCount(RelationRestrictionContext *restrictionContext);
+static List * GenerateAttributeEquivalencesForRelationRestrictions(
+	RelationRestrictionContext *restrictionContext);
+static AttributeEquivalenceClass * AttributeEquivalenceClassForEquivalenceClass(
+	EquivalenceClass *plannerEqClass, RelationRestriction *relationRestriction);
+static void AddToAttributeEquivalenceClass(AttributeEquivalenceClass **
+										   attributeEquivalanceClass,
+										   PlannerInfo *root, Var *varToBeAdded);
+static Var * GetVarFromAssignedParam(List *parentPlannerParamList,
+									 Param *plannerParam);
+static List * GenerateAttributeEquivalencesForJoinRestrictions(JoinRestrictionContext
+															   *joinRestrictionContext);
+static bool AttributeClassContainsAttributeClassMember(AttributeEquivalenceClassMember *
+													   inputMember,
+													   AttributeEquivalenceClass *
+													   attributeEquivalenceClass);
+static List * AddAttributeClassToAttributeClassList(List *attributeEquivalenceList,
+													AttributeEquivalenceClass *
+													attributeEquivalance);
+static bool AttributeEquivalancesAreEqual(AttributeEquivalenceClass *
+										  firstAttributeEquivalance,
+										  AttributeEquivalenceClass *
+										  secondAttributeEquivalance);
+static AttributeEquivalenceClass * GenerateCommonEquivalence(List *
+															 attributeEquivalenceList);
+static void ListConcatUniqueAttributeClassMemberLists(AttributeEquivalenceClass **
+													  firstClass,
+													  AttributeEquivalenceClass *
+													  secondClass);
+
+/*
+ * AllRelationsJoinedOnPartitionKey aims to deduce whether each of the RTE_RELATION
+ * is joined with at least one another RTE_RELATION on their partition keys. If each
+ * RTE_RELATION follows the above rule, we can conclude that all RTE_RELATIONs are
+ * joined on their partition keys.
+ *
+ * The function returns true if all relations are joined on their partition keys.
```

Comment:
Are we actually guaranteeing anywhere that the relevant tables are all "co-partitioned"?